lib/apiservers/engine/backends/container.go

```diff
@@ -1605,6 +1599,62 @@ func (c *Container) containerAttach(name string, ca *backend.ContainerAttachConf
 	return nil
 }
 
+// ContainerRename changes the name of a container, using the oldName
+// to find the container. An error is returned if newName is already
+// reserved.
+func (c *Container) ContainerRename(oldName, newName string) error {
+	defer trace.End(trace.Begin(newName))
+
+	if oldName == "" || newName == "" {
+		err := fmt.Errorf("neither old nor new names may be empty")
+		log.Errorf("%s", err.Error())
+		return err
+	}
+
+	if !utils.RestrictedNamePattern.MatchString(newName) {
+		err := fmt.Errorf("invalid container name (%s), only %s are allowed", newName, utils.RestrictedNameChars)
+		log.Errorf("%s", err.Error())
+		return err
+	}
+
+	// Look up the container name in the metadata cache to get long ID
+	vc := cache.ContainerCache().GetContainer(oldName)
+	if vc == nil {
+		log.Errorf("Container %s not found", oldName)
+		return NotFoundError(oldName)
+	}
+
+	oldName = vc.Name
+	if oldName == newName {
+		err := fmt.Errorf("renaming a container with the same name as its current name")
+		log.Errorf("%s", err.Error())
+		return err
+	}
+
+	if exists := cache.ContainerCache().GetContainer(newName); exists != nil {
+		err := fmt.Errorf("conflict. The name %q is already in use by container %s. You have to remove (or rename) that container to be able to re use that name.", newName, exists.ContainerID)
+		log.Errorf("%s", err.Error())
+		return derr.NewRequestConflictError(err)
+	}
+
+	if err := c.containerProxy.Rename(vc, newName); err != nil {
+		log.Errorf("Rename error: %s", err)
+		return err
+	}
+
+	// update containerCache
+	if err := cache.ContainerCache().UpdateContainerName(oldName, newName); err != nil {
```

Comment:
You're going to get a race condition here.  If you check above for availability, do the the reconfigure, then update the cache, another operation can come in during that time and take your name "slot".  It's better to "reserve" your name slot.  Perform the reconfigure.  Then update the cache.  To reserve the name slot, the easiest way to do that is to add the container with the new name while keeping the old name in the cache.  Then delete the old name after the reconfigure.